Fix for specifying a directory in a package when deploying from windows

w00key opened the pull request:
When using specifying a directory in the package function windows paths becomes a mix of \ and /.

This commit changes the way the Finder class retrieves the path.

Since `SplFileInfo::getPathname` returns a mixture of paths on windows and creating zips on windows really likes / . I replaced the call to `getPathname()` with `$file->getPath() . '/'. $file->getFilename()` both functions on `SplFileInfo`

This means the \ that gets added between the folder and file in question is always a /.

This change may need testing on other platforms but should be okay.

aarondfrancis commented:
Thanks for continuing to run this down. What a pain!

The only thing I can't quite figure out is if this solves for directory separators in the path? I see that the last one between the path and the directory is replaced, but what about ones in the path? 

I'm wondering if we should do something like

```php
str_replace(DIRECTORY_SEPARATOR, "/", $file->getPathname())
```

that would ensure we get them all, right? 

Lemme know what you think about that!

w00key commented:
I'm not sure where is the best place for some of these fixes, this first PR looked like the main place however didn't resolve folders, because it looks like the `Finder` class retrieves the path differently when dealing with folders.

Finder path for the file seemed to be `\folders\folders\file.js`, the folder PR weirdly is happy I think as long as its `/file.js` not `\file.js` presumably to do with the removal of the base path at some point as others have suggested.

So if I'm understanding you correctly, the downstream fix of detecting any `\` and fixing for windows could work, I'll test it out.

w00key commented:
ok did a quick test, fixing all `\` to `/`  at the `Finder::yieldSelectedFiles` level breaks the path for windows to retrieve the file, same if its applied in `Package::prependBasePath` from the original fix.

aarondfrancis commented:
I wonder if we should just do the directory replacement as the file is going [into the zip](https://github.com/hammerstonedev/sidecar/blob/main/src/Package.php#L345-L367)?

That way PHP can always read from the filesystem, and Lambda always ends up with `/`.

Would that solve it in every case?

w00key commented:
Adding your suggested fix to `Package::upload` does also resolve the issue.

```php
$zip->addFileFromPath(
    str_replace(DIRECTORY_SEPARATOR, '/', $this->removeBasePath($file)), $file, $options
);
```
Thats for the `foreach ($this->files() as $file) { ` loop.

Likely need to also change the other `$zip->addFileFromPath` calls in upload, this resolves for using a folder and a path.

You can also revert the previous DIRECTORY_SEPARATOR change along with not using this suggested PR.

I can make a new PR if you want.

aarondfrancis commented:
Nice! I think this is the way to go. Maybe we add a new method, something like `normalizeSeparators` or something? And use it wherever we add to the zip?

w00key commented:
Ok think I've updated this PR, added `normalizeSeparators` and its called in `upload` when adding to the zip. Still new to making PRs but glad to help out.

aarondfrancis commented:
Thanks for seeing this through! 💪
